p4-16/psa/psa.p4

```diff
@@ -776,60 +719,308 @@ extern Digest<T> {
+// copies after this control is executed.
+control CloneE2EPacker<H, M, CE2EM>(
+    in H hdr,  // TBD: Should this be here?
+    in M meta,
+    out CE2EM clone_e2e_meta);
+
+// The same notes about multicast copies applies for the
+// CloneE2EUnpacker control as for the NormalUnpacker control.  See
+// the documentation of the NormalUnpacker for details.
+control CloneE2EUnpacker<M, CE2EM>(
+    in CE2EM clone_e2e_meta,
+    inout M meta);
+
+// Default "empty" or "no op" implementations of packers and
+// unpackers, ready for use for PSA programs that do not need to
+// preserve metadata for those kinds of packets, or do not exercise
+// those packet paths at all.
+
+struct psa_empty_t {
+}
+
+control EmptyNewPacketMetadataInitializer<M>(
+    in PortId_t ingress_port,
+    inout M meta)
+{
+    apply { }
+}
+
+control EmptyNormalPacker<H, M>(
+    in H hdr,  // TBD: Should this be here?
+    in M meta,
+    out psa_empty_t normal_meta)
+{
+    apply { }
+}
+
+control EmptyNormalUnpacker<M>(
+    in psa_empty_t normal_meta,
+    inout M meta)
+{
+    apply { }
+}
+
+control EmptyResubmitPacker<H, M>(
+    in H hdr,  // TBD: Should this be here?
+    in M meta,
+    out psa_empty_t resubmit_meta)
+{
+    apply { }
+}
+
+control EmptyResubmitUnpacker<M>(
+    in psa_empty_t resubmit_meta,
+    inout M user_meta)
+{
+    apply { }
+}
+
+control EmptyRecirculatePacker<H, M>(
+    in H hdr,  // TBD: Should this be here?
+    in M user_meta,
+    out psa_empty_t recirculate_meta)
+{
+    apply { }
+}
+
+control EmptyRecirculateUnpacker<M>(
+    in psa_empty_t recirculate_meta,
+    inout M user_meta)
+{
+    apply { }
+}
+
+control EmptyCloneI2EPacker<H, M>(
+    in H hdr,  // TBD: Should this be here?
+    in M meta,
+    out psa_empty_t clone_i2e_meta)
+{
+    apply { }
+}
+
+control EmptyCloneI2EUnpacker<M>(
+    in psa_empty_t clone_i2e_meta,
+    inout M meta)
+{
+    apply { }
+}
+
+control EmptyCloneE2EPacker<H, M>(
+    in H hdr,  // TBD: Should this be here?
+    in M meta,
+    out psa_empty_t clone_e2e_meta)
+{
+    apply { }
+}
+
+control EmptyCloneE2EUnpacker<M>(
+    in psa_empty_t clone_e2e_meta,
+    inout M meta)
+{
+    apply { }
+}
+
+// TBD: Perhaps there is a way to assign a default parameter value of
+// rsu of EmptyResubmitUnpacker, and similarly for the other packers
+// and unpackers?  The straightforward P4_16 default parameter syntax
+// did not work in my experiments, which may be a bug in p4c.
 
 package IngressPipeline<IH, IM, NM, CI2EM, RESUBM, RECIRCM>(
-    IngressParser<IH, IM, RESUBM, RECIRCM> ip,
+    IngressParser<IH, IM> ip,
     Ingress<IH, IM> ig,
-    IngressDeparser<IH, IM, CI2EM, RESUBM, NM> id);
+    IngressDeparser<IH, IM> id,
+    NewPacketMetadataInitializer<IM> npi,
+    ResubmitUnpacker<IM, RESUBM> rsu,
+    RecirculateUnpacker<IM, RECIRCM> rcu,
+    NormalPacker<IH, IM, NM> np,
+    ResubmitPacker<IH, IM, RESUBM> rsp,
+    CloneI2EPacker<IH, IM, CI2EM> ci2ep);
```

Comment:
I like the direction. The one improvement that I would suggest is to make these additional controls optional, or simply provide a default empty control implementation, such that a P4 program that does not need them, does not refer to them at all. I believe the named arguments support that we currently have in the compiler is almost there.

BTW, I assume this is what we'll discuss later today.

Reply:
Yes, it is what I planned to talk with you about later today.

There are default "empty" packer/unpacker implementations in the propose psa.p4 include file changes, and those are used in the example programs I updated.

I did try to make these additional package variables optional, but p4c today didn't like what I tried.  Either it has a bug in handling default parameter values to packages, or my experiment was wrong somehow.  In writing this, I just realized that I neglected to use an `@optional` annotation on them, which might be why it didn't work for me.  I will try that.

Reply:
One aspect of this that is perhaps more minor: it should be easier to implement the compiler for this approach.

Why?  Because the PSA v1.1 approach requires checking that all assignments to output parameters like recirculate_meta are inside of an `if` statement with the correct condition.  With this approach, _everything_ in the RecirculatePacker is effectively inside of such an `if` statement, automatically, and it is impossible inside of the RecirculatePacker control to write code that affects any other packet paths, e.g. normal, clone, or resubmit.

Reply:
Created an issue for p4c asking if it is possible to have a package declaration with default values for its parameters, and then instantiate that package.  I could not figure out how to get that to work: https://github.com/p4lang/p4c/issues/1914